documentation/guide-api-first.asciidoc

```diff
@@ -0,0 +1,144 @@
+:toc: macro
+toc::[]
+
+= API first development guide
+
+Cloud native promotes the use of microservices, which are loosely coupled and self-contained. These services communicate with each other using a well-defined interface or API, and no consumer needs to be aware of any implementation details of the provider. There could even be multiple providers of the same API, or we can decide to swap existing implementation for a new one, without disrupting our clients - all because we adhere to a well defined API. 
+This guide focuses on HTTP interfaces following RESTful design principles. 
+
+== API first strategy 
+API first strategy treats APIs as first class citizens in the delivery model.  
+The APIs are modeled/designed first (usually as link:guide-openapi.asciidoc[OpenAPI specification]), often in a collaborative process between providers and consumers, and only once the APIs are defined do we start with development of the provider service. 
+This requires a bit more time upfront, but also provides opportunity to think about the behaviour of the system before it gets implemented. Several other advatages of this approach:
+
+ * provider and client implementation can run in parallel - we have the contract, client does not need to wait for provider to finish implementation
+ * support for automation and great ecosystem - we can easily generate stubs and clients for most languages/frameworks, generate documentation and API catalogues using one of the many opensource and commercial tools
+
+== API first provider 
+
+After careful planning, we defined our future API. Now we would like to implement a provider of this API.
+We could manually create all the link:guide-rest.asciidoc#jax-rs[JAX-RS] endpoints based on the schema and then test whether the actually provided interface conforms with the API schema. However, it is an error prone and laborious process, especially if the API gets big.
+Luckily, we can use great open source tooling to generate JAX-RS interfaces that conform to 100% with the schema, and then we simply implement them.
+
+There are many open source tools, that allow code generation from OpenAPI schema files, for example https://openapi-generator.tech/[OpenAPI Generator] or https://github.com/swagger-api/swagger-codegen[Swagger codegen]. Both are based on Java and even provide maven plugins to integrate them in our build.
```

Comment:
There are many open source tools(no comma) that allow code generation from OpenAPI schema files, for example**,** [OpenAPI Generator] or [Swagger codegen].